### PR TITLE
Add fast traceback utilities

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107439
* #107471
* #107388
* __->__ #107358
* #107438

This adds some utilities for conveniently working with fast combined CapturedTraceback from Python. The main goal of these utilities is to make it easier for people to use CapturedTraceback as a drop-in replacement for `traceback.extract_stack`, which is 20x slower than CapturedTraceback.

I port symbolic shapes to use the new CapturedTraceback code, to validate that the APIs work and are useful.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

### CODE REVIEW

File: torch/_dynamo/utils.py [hunk 1]
```diff
@@ -1529,16 +1530,6 @@ def tensor_always_has_static_shape(
...
 
 
-class LazyString:
-    def __init__(self, func, *args, **kwargs):
-        self.func = func
-        self.args = args
-        self.kwargs = kwargs
-
-    def __str__(self):
-        return self.func(*self.args, **self.kwargs)
```
Comment (ezyang) [inline torch/_dynamo/utils.py:1539]: Just a small cleanup; this got moved to torch._logging. I actually didn't need it in this PR but I kept the cleanup. If you REALLY care I can split it out.

File: torch/_guards.py [hunk 2]
```diff
@@ -144,6 +144,11 @@ class Guard:
     obj_weakref: Optional[object] = None
     guarded_class_weakref: Optional[type] = None
 
+    stack = None
+
+    def __post_init__(self):
+        self.stack = CapturedTraceback.extract(skip=2)
```
Comment (ezyang) [inline torch/_guards.py:150]: Not used yet but I plan to use it later.

File: torch/fx/experimental/symbolic_shapes.py [hunk 3]
```diff
@@ -1059,11 +1058,6 @@ def error():
     raise AssertionError("shouldn't be hit")
 
 
-def get_debugging_stack(num_frames_to_cut=2):
-    # cut this frame and the caller's frame by default
-    return ''.join(traceback.format_list(traceback.extract_stack()[:-num_frames_to_cut]))
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:1064]: Dead code.

File: torch/fx/experimental/symbolic_shapes.py [hunk 4]
```diff
@@ -3338,11 +3332,22 @@ def _check_frozen(self, expr, concrete_val):
...
+                    if frame.f_code.co_filename not in uninteresting_files():
+                        fsummary = traceback.FrameSummary(
+                            frame.f_code.co_filename,
+                            frame.f_lineno,
+                            frame.f_code.co_name,
+                        )
+                        break
+                    frame = frame.f_back
+            finally:
+                del frame
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:3350]: I don't have the formatted tb handy anymore, so I reimplement the old code by manually traversing over the frame. This should be faster!

**Review by zdevito — APPROVED** [review]
Use of CapturedTraceback apis looks good to me. Having an API that mimics Python's traceback API will be useful, especially when we want TorchScript/inductor stacks to be readable. Didn't review the use of the new API in symbolic shapes closely because I am not familiar with that code

**Comment by ezyang** [comment]
The symbolic shapes stuff is now in its own PR https://github.com/pytorch/pytorch/pull/107439 because apparently it causes a memory leak in cudagraph trees.

File: torch/utils/_traceback.py [hunk 5]
```diff
@@ -150,3 +151,103 @@ def format_traceback_short(tb):
...
+
+    def __init__(self, tb, skip=0):
+        self.tb = tb
+        self.skip = skip
+        # Cached StackSummary; this mostly exists so CapturedTraceback
+        # can be reliably pickled and then printed later
+        self._summary = None
+
+    def summary(self):
+        import torch._C._profiler
```
Comment (albanD) [inline torch/utils/_traceback.py:166]: Isn't that done by default? :o

**Review by albanD — APPROVED** [review]
SGTM

**Pull request closed without merging**